[docs] Describe how to mock part of response headers (#6089)
DevExpress/testcafe: docs/articles/documentation/reference/test-api/requesthook/README.md

    throw new Error('Not implemented');
}
```

> ### Mock Response Headers

Reviewer: [can be ignored] I'm not quite sure, but I feel that the "stubbing" term is more suitable here. From [Mocks Aren't Stubs](https://martinfowler.com/articles/mocksArentStubs.html):
> Mocks are what we are talking about here: objects pre-programmed with expectations which form a specification of the calls they are expected to receive. 

> Stubs provide canned answers to calls made during the test, usually not responding at all to anything outside what's programmed in for the test.

Author: You are right, this is indeed a stub, not a mock

Before it passes an incoming response to the application, `RequestHook` can modify the response headers. This allows you to mock part of the response.

> > Important! You can not modify the body of the request or its headers inside a `RequestHook.onResponse()` call. If you edit properties of `event.headers` or `event.body` inside `onResponse()`, TestCafe ignores these changes.

Reviewer: I think this note would be more useful in the "Response Handler" section above.

You can use the `RequestHook._onConfigureResponse()` method to substitute response headers, effectively mocking part of the response.

The example below demonstrates a `RequestHook` that sets a `foo:bar` header on the response and removes the `x-header` from the response.

```js
class HeadersHook extends RequestHook {
  constructor (requestFilterRules) {
    super(requestFilterRules);
  }

  async onRequest () { }
  async onResponse () { }

  _onConfigureResponse (event) {
    super._onConfigureResponse(event);

    event.setHeader('foo', 'bar');
    event.removeHeader('x-header');
  }
}

> ```

Reviewer: Let's use more real-world header names:
```js
class HeadersHook extends RequestHook {
    constructor (requestFilterRules) {
        super(requestFilterRules);
    }

    async onRequest () { }
    async onResponse () { }

    _onConfigureResponse (event) {
        super._onConfigureResponse(event);

        event.setHeader('x-token', 'token-123');
        event.removeHeader('x-frame-options');
    }
}
```

    throw new Error('Not implemented');
}
```

{% include notes-warnings/requesthook-onresponse-no-stub-headers.md href='#stub-response-headers' %}

> ### Stub Response Headers

Reviewer 2: Agreed

> ### Stub Response Headers

Author: > Change/Delete Response Headers

> Before it passes an incoming response to the application, `RequestHook` can change the response headers.

Reviewer 3: `Before it passes an incoming response to the application` - what  is it?

> effectively mocking part of the response.
You don't mock part of the response - you change the existing response.

I propose to write the main idea at once:
```js
You can use the `RequestHook._onConfigureResponse()` method to set the header value or remove existing header.
```

Author: See [comment](https://github.com/DevExpress/testcafe/pull/6089#discussion_r603547525)

> The `RequestHook._onConfigureResponse()` method sets and removes headers, effectively mocking part of the response.

Reviewer 3: > The `RequestHook._onConfigureResponse()` method sets and removes headers, effectively mocking part of the response.

This method do nothing. You can inside this method change headers.
```js
_onConfigureResponse (event) {
    super._onConfigureResponse(event);
    event.setHeader('x-token', 'token-123');
    event.removeHeader('x-frame-options');
  }
```

> effectively mocking part of the response.

The `mock` means `create something synthetic instead of real`. 
The HTTP response contains the statusCode, headers and body.
You can modify the part of the existing response (only headers).

Author: > You can set or remove response headers in a `RequestHook._onConfigureResponse()` call.

> You can use the `RequestHook._onConfigureResponse()` method to set the header value or remove existing header, effectively mocking part of the response.

Reviewer 2: My point was: I - the user - do not care that I can use X for Y purpose, unless: a) I already use X or b) it's the best way to achieve Y.

To give you an example: I can - in theory - smash roaches with an antique clock. However, if you say to me that "You can use an antique clock to smash roaches" I'd be confused because a) I don't have an antique clock and b) It sounds suboptimal c) I don't think I have a roach problem 

"You can use..." sounds like: a) you can do other things with the onConfigureResponse method. *If you can - please describe them* b) you can mock request headers with other methods. *If you can - mention them* c) I already want to change headers *Why would I want to change headers?*

```suggestion
The `RequestHook._onConfigureResponse()` method sets and remove headers, effectively mocking part of the response.
```

As of now, I'm not 100% sure this file *as a whole* - not your fault - does a good job of explaining what `RequestHooks`are, and why I would ever need one... 

We should always try to imagine someone reading everything out of context - as if they followed an anchor link on Google - and try to give them a complete enough picture

> You can use the `RequestHook._onConfigureResponse()` method to substitute response headers, effectively mocking part of the response.

Reviewer 2: How?

Author: > You can use the `RequestHook._onConfigureResponse()` method to set the header value or remove existing header, effectively mocking part of the response.
> 
> To change or delete response headers, use `event.setHeaders` and `event.removeHeader` in the `__onConfigureResponse` call.

> To change response headers, use `event.setHeaders` and `event.removeHeader` methods in the `__onConfigureResponse` call.

Reviewer 4: ```suggestion
To change response headers, use the `event.setHeaders` and `event.removeHeader` methods in the `__onConfigureResponse` call.
```

> To change response headers, use the `event.setHeaders` and `event.removeHeader` methods in the `__onConfigureResponse` call.

Reviewer: ```suggestion
To change response headers, use the `event.setHeader` and `event.removeHeader` methods in the `__onConfigureResponse` call.
```

> > Important! The `RequestHook._onConfigureResponse()` method is private. Exercise caution when you implement this method in your request hooks.

Reviewer 2: why `implement`? it is my understanding that this method works out of the box...
also, `exercise caution` is unnecessarily formal and old-fashioned
```suggestion
> Important! The `RequestHook._onConfigureResponse()` method is private. Be cautious when you add this method to your request hooks.
```

> > Important! The `RequestHook._onConfigureResponse()` method is private. Be cautious when you add this method to your request hooks.

Reviewer 3: More technically correct:

`Be cautious when you override this method to your request hooks.`

> > Important! The `RequestHook._onConfigureResponse()` method is private. Be cautious when you add this method to your request hooks.

Reviewer 4: `Be cautious` Why should you be cautious?

> > Important! The `RequestHook._onConfigureResponse()` method is not part of TestCafe public API. As a private method, it can be removed in the upcoming releases without prior warning. Be cautious when you override this method in your request hooks.

Reviewer 2: unnecessary definite article

```suggestion
> Important! The `RequestHook._onConfigureResponse()` method is not part of TestCafe public API. As a private method, it can be removed in upcoming releases without prior warning. Be cautious when you override this method in your request hooks.
```

> > Important! The `RequestHook._onConfigureResponse()` method is not part of TestCafe public API. As a private method, it can be removed or deprecated in the upcoming releases without prior warning. Be cautious when you add this method to your request hooks.

Author: ```suggestion
> Important! The `RequestHook._onConfigureResponse()` method is not part of TestCafe public API. As a private method, it can be removed or deprecated in the upcoming releases without prior warning. Be cautious when you override this method in your request hooks.
```

The example below demonstrates a `RequestHook` that sets a `x-token:token-123` header on the response and removes the `x-frame-options` header from the response.

```js
class HeadersHook extends RequestHook {
  constructor (requestFilterRules) {
    super(requestFilterRules);
  }

  async onRequest () { }
  async onResponse () { }

  _onConfigureResponse (event) {
    super._onConfigureResponse(event);

    event.setHeader('x-token', 'token-123');
    event.removeHeader('x-frame-options');
  }
}
```

> You can not modify the response body with `RequestHook._onConfigureResponse()`.

Reviewer 3: You cannot modify the response body. There is no such functionality.

Reviewer 2: That's exactly what it says :D
